### PR TITLE
9.0 account chart update migration

### PR DESCRIPTION
**Pull request by jbaudoux**
Replaces PR https://github.com/OCA/account-financial-tools/pull/381

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Why this new PR?

**Comment by pedrobaeza** [comment]
Can you please answer why you open a new PR?

**Comment by jbaudoux** [comment]
@pedrobaeza The PR #381 requires to be fixed. How this could have been done other than creating a new branch with the fixes and a new PR?

**Comment by pedrobaeza** [comment]
Well, first asking the contributor to fix it, or making a PR over his branch. Anyway, I see that you have respected his commit history, so let's continue with this one.

**Comment by jbaudoux** [comment]
I'm still making some more tests because tax name was not shown in wizard and tax accounts where not loaded.

**Comment by pedrobaeza** [comment]
Perfect, let me know when I can review it

**Comment by jbaudoux** [comment]
Any good reason why find_tax_by_templates tries to match an existing tax based on its description? The description is what is displayed on the invoice.
For instance, I have multiple fixed amount taxes (and so with different names to be able to select it properly) that are all having the same description. The description is the same because it's all the same tax and there is no reason to display it differently on the invoice whatever the amount is.

**Comment by pedrobaeza** [comment]
There are some localizations that only fill that field. Maybe for v9 the situation has been normalized. What is the field name which you want to remove? description or name?

**Comment by jbaudoux** [comment]
@pedrobaeza No improvement on this in 9.0, so I leave it like that. I'll manage to have different description on those taxes.

I'm done fixing the issues I found. This PR is now ready for review

**Comment by jbaudoux** [comment]
@sbidoul Could you also review this module that allows to install the l10n_belgium PR you reviewed after having the chart of accounts/taxes loaded. Thanks

**Comment by jbaudoux** [comment]
Could this PR be merged?
We are starting porting this PR to version 10

File: account_chart_update/__openerp__.py [hunk 1]
```diff
@@ -1,51 +1,37 @@
...
               "Serv. Tecnol. Avanzados - Pedro M. Baeza, "
               "ACSONE A/V, "
+              "Tecnativa, "
+              "BCIM,"
               "Odoo Community Association (OCA)",
     'website': "http://odoo-community.org",
     'depends': ["account"],
-    'category': "Generic Modules/Accounting",
+    'category': "Accounting & Finance",
     'contributors': [
```
Comment (yajo) [inline account_chart_update/__openerp__.py:24]: Does this key really do anything? 🤔

File: account_chart_update/wizard/wizard_chart_update.py [hunk 2]
```diff
@@ -312,779 +187,513 @@ def onchange_company_id(self):
...
+                         ("type_tax_use", "=", template.type_tax_use)],
+                        limit=1)
+            result |= single
+        return result
+
+    @api.model
+    @tools.ormcache("code")
+    def padded_code(self, code):
+        """Return a right-zero-padded code with the chosen digits."""
+        return "{:0<{}}".format(code, self.code_digits)
```
Comment (yajo) [inline account_chart_update/wizard/wizard_chart_update.py]: Only unicode strings are allowed to be used with `.format` (per OCA standard).

**Review by yajo — APPROVED** [review]
Code review looks good, still lacking tests sadly.

Did not check runbot since it is unavailable ATM.

**Comment by pedrobaeza** [comment]
On a tentative update, there's this error:
```
A tax fiscal position could be defined only once time on same taxes.
```
so something fails.

File: account_chart_update/__openerp__.py (hunk 1 above)
Comment (pedrobaeza) [reply account_chart_update/__openerp__.py:24]: Well, it's loaded, but not shown in any part. We can keep it anyway.

File: account_chart_update/wizard/wizard_chart_update.py (hunk 2 above)
Comment (pedrobaeza) [reply account_chart_update/wizard/wizard_chart_update.py]: Uhm, format can be used with both kind of strings, so no real problem I think.

**Comment by pedrobaeza** [comment]
@jbaudoux I have reordered a bit commits. The changes you have made to the fiscal position part are not working, with the message I put the other day. Can you please check?

File: account_chart_update/wizard/wizard_chart_update.py (hunk 2 above)
Comment (yajo) [reply account_chart_update/wizard/wizard_chart_update.py]:
```
>>> "Vivo en {}".format(u"España")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode character u'\xf1' in position 4: ordinal not in range(128)
>>> u"Vivo en {}".format(u"España")
u'Vivo en Espa\xf1a'
```

File: account_chart_update/views/account_config_settings_view.xml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- © 2016 Jairo Llopis <jairo.llopis@tecnativa.com>
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+<openerp>
```
Comment (cubells) [inline account_chart_update/views/account_config_settings_view.xml]: s/openerp/odoo/g

File: account_chart_update/wizard/wizard_chart_update_view.xml [hunk 4]
```diff
@@ -1,156 +1,133 @@
 <?xml version="1.0" encoding="utf-8"?>
+<!-- © 2016 Jairo Llopis <jairo.llopis@tecnativa.com>
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
 <openerp>
```
Comment (cubells) [inline account_chart_update/wizard/wizard_chart_update_view.xml]: s/openerp/odoo/g

**Review by cubells — CHANGES_REQUESTED** [review]
Code review only.

File: account_chart_update/wizard/wizard_chart_update.py (hunk 2 above)
Comment (pedrobaeza) [reply account_chart_update/wizard/wizard_chart_update.py]: But it's not the case here. Anyway, the same happens with `"Vivo en %s" % "España"` if you don't put the `u`.

Comment (yajo) [reply account_chart_update/wizard/wizard_chart_update.py]: You can see discussion on https://github.com/OCA/pylint-odoo/pull/65, I'd still recommend it as a good rule of thumb.

Comment (pedrobaeza) [reply account_chart_update/wizard/wizard_chart_update.py]: OK, let's change it...

**Comment by jbaudoux** [comment]
@pedrobaeza I cannot reproduce your error.

File: account_chart_update/wizard/wizard_chart_update.py [hunk 5]
```diff
@@ -312,779 +187,513 @@ def onchange_company_id(self):
...
+    def find_fp_tax_by_templates(self, templates):
+        result = []
+        for tpl in templates:
+            pos_id = self.find_fp_by_templates(tpl.position_id).id
+            src_id = self.find_tax_by_templates(tpl.tax_src_id).id
+            dest_id = self.find_tax_by_templates(tpl.tax_dest_id).id
+            mapping = self.env["account.fiscal.position.tax"].search([
+                ("position_id", "=", pos_id),
+                ("tax_src_id", "=", src_id),
+                ], limit=1)
```
Comment (carlosdauden) [inline account_chart_update/wizard/wizard_chart_update.py]: Don't works fine if fiscal position map 1 tax with more than 1 tax.

**Review by carlosdauden — CHANGES_REQUESTED** [review]

**Review by pedrobaeza — APPROVED** [review]
I have fixed the fiscal position part, so this one is finally ready to merge.

**Pull request merged**